sbatson5/firestore-jest-mock
# Begin porting to TypeScript

Sorry about the change stats on this PR. About 40k of the line changes are in `package-lock.json`, and most of the files changed are simple renames. Prettier fixed some tables in `README.md`. Should be easy enough to grok otherwise.

# Description

This was a lot simpler than I first assumed, considering `tsc` can totally output to the workspace root so we don't have to break consumers' import paths!

The goal is to simplify the maintenance of `.d.ts` type files, and make eventual features easier to implement (like an eventual ESM build, Vitest support, etc.)

I've moved all source files to the `src/` directory, and set up the build script to output to the same place our source files were before: the workspace root. This is a little messy, but I didn't want to break things by moving output to `dist/` or `lib/`. We may do that later on, but for now, I just want to start using TypeScript directly, rather than juggling type files.

Also, I need Vitest support for other projects, and testing that is _much_ easier if the tests can transform TS sources to ES modules separately (per Vitest's ESM requirement), while building to CJS modules for production.

## Potential Future Work

- Begin combining old `.js` and `.d.ts` sources into new `.ts` sources. (Fun non-breaking change)
- Move build output to `lib/` to clean up the workspace root. (Semver-major change)
- Concurrent ESM and CJS builds. (May require reorganizing exports if we still intend to support deep import paths later on)

## Related issues

Fixes #20.

Addresses #67.

## How to test

`npm t` to run tests.

`npm run build` to try the build.

There should be no runtime changes here, even in import paths.
